File: client/src/models/Battery.js
import { Device } from "./Device";

export class Battery extends Device {
  constructor({
    id,
    derId,
    icon,
    name,
    type,
    category,
    maximumAvailablePower,
    capacity,
    minStateOfCharge,
    maxStateOfCharge,
    createdAt,
    updatedAt,
    isArchived,
  }) {
    super({
      id,
      derId,
      icon,
      name,
      type,
      category,
      maximumAvailablePower,
      createdAt,
      updatedAt,
      isArchived,
    });

    this.capacity = parseFloat(capacity) || undefined;
    this.minStateOfCharge = parseFloat(minStateOfCharge) || undefined;
    this.maxStateOfCharge = parseFloat(maxStateOfCharge) || undefined;
  }
}

File: server/src/main/java/com/schneider/project/model/Battery.java
package com.schneider.project.model;

import com.schneider.project.model.enums.DeviceCategory;
import jakarta.persistence.*;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotNull;
import lombok.*;

@Entity
@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
@ToString
@PrimaryKeyJoinColumn(name = "id")
public class Battery extends Device {
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private DeviceCategory category;

    @PrePersist
    private void prePersist() {
        this.category = DeviceCategory.MIXED;
    }

    @Column(nullable = false)
    @NotNull
    @Min(value = 0, message = "Capacity must be greater than or equal to 0")
    @Max(value = 10, message = "Capacity must be less than or equal to 10")
    private Double capacity;

    @Column(nullable = false)
    @NotNull
    @Min(value = 0, message = "Minimum state of charge must be greater than or equal to 0")
    @Max(value = 20, message = "Minimum state of charge must be less than or equal to 20")
    private Double minStateOfCharge;

    @Column(nullable = false)
    @NotNull
    @Min(value = 90, message = "Maximum state of charge must be greater than or equal to 90")
    @Max(value = 100, message = "Maximum state of charge must be less than or equal to 100")
    private Double maxStateOfCharge;

    @Column(nullable = false)
    @NotNull
    @Min(value = 0, message = "Maximum available power must be greater than or equal to 0")
    @Max(value = 10, message = "Maximum available power must be less than or equal to 10")
    private Double maximumAvailablePower;
}
